# Task 02 - Track B: Advanced DataFrame Operations

**Course:** Database Applications Development  
**Lesson:** 02 - Working with DataFrames (Preparing for SQL)  

---

## Instructions

Complete all exercises in this notebook. Track B includes significantly more practice with multiple conditions, chaining operations, and deeper analysis.

**Submission:**
1. Save this notebook as `dbAppsTask02TrackB.ipynb`
2. Add, commit, and push to your `databaseApplications` repository on GitHub
3. Verify the file appears correctly on GitHub

---

## Setup: Load the Dataset

In [109]:
# Import pandas and load data
import pandas as pd

titanic = pd.read_csv('Titanic Dataset.csv')

# Quick check
print(titanic.shape)
titanic.head()

(1309, 14)


pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

     age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.00      0      0   24160  211.3375       B5        S    2    NaN   
1   0.92      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.00      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

---

## Part 1: Selecting Columns

### Exercise 1.1: Select Single Columns

Select and display the first 10 values:
1. The 'age' column
2. The 'fare' column

In [110]:
# Age column
print("Age column:")
print(titanic['age'].head(10))


Age column:
0    29.00
1     0.92
2     2.00
3    30.00
4    25.00
5    48.00
6    63.00
7    39.00
8    53.00
9    71.00
Name: age, dtype: float64


In [111]:
# Fare column
print("Fare column:")
print(titanic['fare'].head(10))

Fare column:
0    211.3375
1    151.5500
2    151.5500
3    151.5500
4    151.5500
5     26.5500
6     77.9583
7      0.0000
8     51.4792
9     49.5042
Name: fare, dtype: float64


### Exercise 1.2: Select Multiple Columns

Create a DataFrame called `passenger_info` with these columns: 'name', 'sex', 'age', 'survived'

Display the first 10 rows.

In [112]:
# Your code here
passenger_info = titanic [['name', 'sex', 'age', 'survived']]

print(passenger_info.head(10))



                                              name     sex    age  survived
0                    Allen, Miss. Elisabeth Walton  female  29.00         1
1                   Allison, Master. Hudson Trevor    male   0.92         1
2                     Allison, Miss. Helen Loraine  female   2.00         0
3             Allison, Mr. Hudson Joshua Creighton    male  30.00         0
4  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female  25.00         0
5                              Anderson, Mr. Harry    male  48.00         1
6                Andrews, Miss. Kornelia Theodosia  female  63.00         1
7                           Andrews, Mr. Thomas Jr    male  39.00         0
8    Appleton, Mrs. Edward Dale (Charlotte Lamson)  female  53.00         1
9                          Artagaveytia, Mr. Ramon    male  71.00         0


---

## Part 2: Filtering Rows

### Exercise 2.1: Filter by Survival

Create a DataFrame called `survivors` with only passengers who survived.

Print how many survivors there were.

In [113]:
# Your code here
survivors = titanic[titanic['survived'] == 1]

print("Number of survivors:", len(survivors))


Number of survivors: 500


### Exercise 2.2: Filter by Gender

Create a DataFrame called `females` with only female passengers.

Print how many there were.

In [114]:
# Your code here
females = titanic[titanic['sex'] == 'female']

print("Number of female passengers:", len(females))


Number of female passengers: 466


### Exercise 2.3: Filter by Class

Create a DataFrame called `first_class` with only first class passengers.

Print how many there were.

In [115]:
# Your code here
first_class = titanic[titanic['pclass'] == 1]

print("Number of 1st class passengers:", len(first_class))


Number of 1st class passengers: 323


### Exercise 2.4: Filter by Age

Create a DataFrame called `children` with passengers 12 or younger.

Print how many there were.

In [116]:
# Your code here
children = titanic[titanic['age'] <= 12]

print("There were", len(children), "children on the titanic.")


There were 94 children on the titanic.


---

## Part 3: Sorting Data

### Exercise 3.1: Sort by Age

Sort passengers by age and display name and age for the 10 youngest.

In [117]:
# Your code here
titanic[['name','age']].sort_values(by = 'age').head(10)

name   age
763   Dean, Miss. Elizabeth Gladys "Millvina"  0.17
747   Danbom, Master. Gilbert Sigvard Emanuel  0.33
1240          Thomas, Master. Assad Alexander  0.42
427                 Hamalainen, Master. Viljo  0.67
658             Baclini, Miss. Helene Barbara  0.75
657                    Baclini, Miss. Eugenie  0.75
1111           Peacock, Master. Alfred Edward  0.75
611                 Aks, Master. Philip Frank  0.83
359             Caldwell, Master. Alden Gates  0.83
548           Richards, Master. George Sibley  0.83

### Exercise 3.2: Sort by Fare

Sort passengers by fare (highest first) and display name and fare for the top 10.

In [118]:
# Your code here
titanic[['name', 'fare']].sort_values(by = 'fare', ascending = False).head(10)


name      fare
302                                   Ward, Miss. Anna  512.3292
49                  Cardeza, Mr. Thomas Drake Martinez  512.3292
50   Cardeza, Mrs. James Warburton Martinez (Charlo...  512.3292
183                             Lesurer, Mr. Gustave J  512.3292
116                Fortune, Mrs. Mark (Mary McDougald)  263.0000
111                     Fortune, Miss. Alice Elizabeth  263.0000
115                                  Fortune, Mr. Mark  263.0000
114                     Fortune, Mr. Charles Alexander  263.0000
113                         Fortune, Miss. Mabel Helen  263.0000
112                         Fortune, Miss. Ethel Flora  263.0000

---

## Part 4: Statistics

### Exercise 4.1: Overall Statistics

Calculate and print:
1. Average age
2. Average fare
3. Maximum fare
4. Minimum age

In [119]:
# Your code here
average_age = titanic['age'].mean()
print(f"1. Average age: {average_age:.0f} years")

average_fare = titanic['fare'].mean()
print(f"2. Average fare: ${average_fare:.2f}")

max_fare = titanic['fare'].max()
print(f"3. Maximum fare: ${max_fare:.2f}")

min_age = titanic['age'].min()
print(f"4. Minimum age: {min_age:.2f} years")


1. Average age: 30 years
2. Average fare: $33.30
3. Maximum fare: $512.33
4. Minimum age: 0.17 years


### Exercise 4.2: Statistics on Filtered Data

Calculate:
1. Average age of survivors
2. Average fare of first class passengers

In [120]:
# Your code here
avgAgeSurvived = titanic[titanic['survived'] == 1]['age'].mean()
print(f"1. Average age of survivors: {avgAgeSurvived:.0f} years")

avgFare = titanic[titanic['pclass'] == 1]['fare'].mean()
print(f"2. Average fare of 1st class: ${avgFare:.2f}")


1. Average age of survivors: 29 years
2. Average fare of 1st class: $87.51


---

## Part 5: Value Counts

### Exercise 5.1: Count Values

Use `.value_counts()` to find:
1. How many in each class
2. How many male vs female
3. How many survived vs died

In [121]:
# Class counts
classMem = titanic['pclass'].value_counts()
print(f"1. Counts of people in each class: {classMem}")

1. Counts of people in each class: pclass
3    709
1    323
2    277
Name: count, dtype: int64


In [122]:
# Gender counts
maleFem = titanic['sex'].value_counts()
print(f"2. Count of males and females: {maleFem}")

2. Count of males and females: sex
male      843
female    466
Name: count, dtype: int64


In [123]:
# Survival counts
survive = titanic['survived'].value_counts()
print(f"3. Count of survivors vs deceased (1 = Survived 2 = Died): {survive}")

3. Count of survivors vs deceased (1 = Survived 2 = Died): survived
0    809
1    500
Name: count, dtype: int64


---

## Part 6: Multiple Condition Filtering (Track B)

Combine conditions using & (AND) and | (OR). Remember: each condition needs parentheses!

### Example: Combining Conditions

In [124]:
# AND (&) - Both conditions must be True
# Find male survivors
male_survivors = titanic[
    (titanic['sex'] == 'male') & 
    (titanic['survived'] == 1)
]
print(f"Male survivors: {len(male_survivors)}")

# OR (|) - At least one condition must be True
# Find first or second class passengers
upper_class = titanic[
    (titanic['pclass'] == 1) | 
    (titanic['pclass'] == 2)
]
print(f"Upper class passengers: {len(upper_class)}")

# CRITICAL: Each condition MUST have parentheses around it!

Male survivors: 161
Upper class passengers: 600


### Exercise 6.1: Female Survivors

Find all female passengers who survived.

How many were there?

In [125]:
# Your code here
# Hint: (titanic['sex'] == 'female') & (titanic['survived'] == 1)
female_survivors = titanic[
    (titanic['sex'] == 'female') &
    (titanic['survived'] == 1)
]
print(f"There were {len(female_survivors)} female survivors.")


There were 339 female survivors.


### Exercise 6.2: Upper Class Passengers

Find passengers in either first OR second class.

How many were there?

In [126]:
# Your code here
# Hint: (titanic['pclass'] == 1) | (titanic['pclass'] == 2)
upp_class = titanic[
    (titanic['pclass'] == 1) |
    (titanic['pclass'] == 2)
]

print(f"Upper class passengers: {len(upp_class)}")


Upper class passengers: 600


### Exercise 6.3: Male First Class Survivors

Find male passengers in first class who survived.

Display name, age, and fare for all of them.

In [127]:
# Your code here (three conditions with &)
male_first = titanic[
    (titanic['pclass'] == 1) &
    (titanic['sex'] == "male") &
    (titanic['survived'] == 1)
]

print(f"Male 1st class survivors: {len(male_first)}")
print(male_first[['name', 'age', 'fare']])


Male 1st class survivors: 61
                                     name    age      fare
1          Allison, Master. Hudson Trevor   0.92  151.5500
5                     Anderson, Mr. Harry  48.00   26.5500
14   Barkworth, Mr. Algernon Henry Wilson  80.00   30.0000
20          Beckwith, Mr. Richard Leonard  37.00   52.5542
22                  Behr, Mr. Karl Howell  26.00   30.0000
..                                    ...    ...       ...
292              Taylor, Mr. Elmer Zebley  48.00   52.0000
295           Thayer, Mr. John Borland Jr  17.00  110.8833
298       Tucker, Mr. Gilbert Milligan Jr  31.00   28.5375
317       Williams, Mr. Richard Norris II  21.00   61.3792
320                     Woolner, Mr. Hugh    NaN   35.5000

[61 rows x 3 columns]


### Exercise 6.4: Children Who Survived

Find passengers who were 12 or younger AND survived.

How many children survived?

In [128]:
# Your code here
child_survive = titanic[
    (titanic['age'] <= 12) &
    (titanic['survived'] == 1)
]

print(f"Survivors thate were 12 and younger: {len(child_survive)}")


Survivors thate were 12 and younger: 54


### Exercise 6.5: Expensive Tickets in Third Class

Find third class passengers who paid more than $20 for their fare.

How many were there? (This seems unusual - third class was cheap!)

In [129]:
# Your code here
third_classFareG20 = titanic [
    (titanic['pclass'] == 3) &
    (titanic['fare'] > 20)
]

print(f"{len(third_classFareG20)} people in 3rd class paid more than $20, that's outrageous!!")

116 people in 3rd class paid more than $20, that's outrageous!!


### Exercise 6.6: Complex Filter

Find passengers who meet ALL these conditions:
- Female
- Age between 20 and 40 (inclusive)
- Paid more than $30 fare

Display name, age, fare for the first 10.

In [130]:
# Your code here (four conditions!)
filter = titanic[
    (titanic['sex'] == "female") &
    (titanic['age'] <= 40) &
    (titanic['age'] >= 20) &
    (titanic['fare'] > 30)
]
print(f"There were {len(filter)} women.")
print(filter[['name', 'age', 'fare']].head(10))

There were 77 women.
                                               name   age      fare
0                     Allen, Miss. Elisabeth Walton  29.0  211.3375
4   Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  25.0  151.5500
12                    Aubart, Mme. Leontine Pauline  24.0   69.3000
13                     Barber, Miss. Ellen "Nellie"  26.0   78.8500
18                            Bazzani, Miss. Albina  32.0   76.2917
24                                Bird, Miss. Ellen  29.0  221.7792
28                           Bissette, Miss. Amelia  35.0  135.6333
32                          Bonnell, Miss. Caroline  30.0  164.8667
36                      Bowerman, Miss. Elsie Edith  22.0   55.0000
57        Carter, Mrs. William Ernest (Lucile Polk)  36.0  120.0000


---

## Part 7: Advanced Sorting (Track B)

### Example: Sort by Multiple Columns

In [131]:
# Sort by class (ascending), then by fare within each class (descending)
sorted_df = titanic.sort_values(
    ['pclass', 'fare'],           # List of columns to sort by
    ascending=[True, False]        # True for ascending, False for descending
)

# Display name, class, and fare for first 20
print(sorted_df[['name', 'pclass', 'fare']].head(20))

# This groups by class first, then sorts by fare within each class

                                                  name  pclass      fare
49                  Cardeza, Mr. Thomas Drake Martinez       1  512.3292
50   Cardeza, Mrs. James Warburton Martinez (Charlo...       1  512.3292
183                             Lesurer, Mr. Gustave J       1  512.3292
302                                   Ward, Miss. Anna       1  512.3292
111                     Fortune, Miss. Alice Elizabeth       1  263.0000
112                         Fortune, Miss. Ethel Flora       1  263.0000
113                         Fortune, Miss. Mabel Helen       1  263.0000
114                     Fortune, Mr. Charles Alexander       1  263.0000
115                                  Fortune, Mr. Mark       1  263.0000
116                Fortune, Mrs. Mark (Mary McDougald)       1  263.0000
35                            Bowen, Miss. Grace Scott       1  262.3750
66                         Chaudanson, Miss. Victorine       1  262.3750
249                        Ryerson, Master. John Bo

### Exercise 7.1: Sort by Multiple Columns

Sort passengers by class (ascending), then by fare within each class (descending).

Display name, pclass, fare for the first 20 rows.

In [132]:
# Your code here
# Hint: sort_values(['pclass', 'fare'], ascending=[True, False])
sortedByClass = titanic.sort_values(
        ['pclass'],
        ascending = True
)

print(sortedByClass[['name', 'pclass', 'fare']].head(20))


                                                 name  pclass      fare
31                                   Blank, Mr. Henry       1   31.0000
32                            Bonnell, Miss. Caroline       1  164.8667
20                      Beckwith, Mr. Richard Leonard       1   52.5542
0                       Allen, Miss. Elisabeth Walton       1  211.3375
24                                  Bird, Miss. Ellen       1  221.7792
30                       Blackwell, Mr. Stephen Weart       1   35.5000
29          Bjornstrom-Steffansson, Mr. Mauritz Hakan       1   26.5500
12                      Aubart, Mme. Leontine Pauline       1   69.3000
11  Astor, Mrs. John Jacob (Madeleine Talmadge Force)       1  227.5250
10                             Astor, Col. John Jacob       1  227.5250
9                             Artagaveytia, Mr. Ramon       1   49.5042
8       Appleton, Mrs. Edward Dale (Charlotte Lamson)       1   51.4792
25                                Birnbaum, Mr. Jakob       1   

### Exercise 7.2: Who Paid the Least?

Find the 10 passengers who paid the lowest fares.

Display name, pclass, and fare.

In [133]:
# Your code here
paidLeast = titanic.sort_values(
        ['fare'],
        ascending = True
)

print(sortedByClass[['fare', 'name', 'pclass']].head(10))


        fare                                               name  pclass
31   31.0000                                   Blank, Mr. Henry       1
32  164.8667                            Bonnell, Miss. Caroline       1
20   52.5542                      Beckwith, Mr. Richard Leonard       1
0   211.3375                      Allen, Miss. Elisabeth Walton       1
24  221.7792                                  Bird, Miss. Ellen       1
30   35.5000                       Blackwell, Mr. Stephen Weart       1
29   26.5500          Bjornstrom-Steffansson, Mr. Mauritz Hakan       1
12   69.3000                      Aubart, Mme. Leontine Pauline       1
11  227.5250  Astor, Mrs. John Jacob (Madeleine Talmadge Force)       1
10  227.5250                             Astor, Col. John Jacob       1


### Exercise 7.3: Oldest in Each Class

Sort by class (ascending) and age (descending).

This groups each class together with oldest first. Display name, pclass, age for first 30.

In [134]:
# Your code here
result = titanic.sort_values(['pclass', 'age'], ascending=[True, False])
print(result[['name', 'pclass', 'age']].head(30))

                                                  name  pclass   age
14                Barkworth, Mr. Algernon Henry Wilson       1  80.0
61   Cavendish, Mrs. Tyrell William (Julia Florence...       1  76.0
9                              Artagaveytia, Mr. Ramon       1  71.0
135                          Goldschmidt, Mr. George B       1  71.0
81                        Crosby, Capt. Edward Gifford       1  70.0
285                                 Straus, Mr. Isidor       1  67.0
205                          Millet, Mr. Francis Davis       1  65.0
221                     Ostby, Mr. Engelhart Cornelius       1  65.0
78   Compton, Mrs. Alexander Taylor (Mary Eliza Ing...       1  64.0
83   Crosby, Mrs. Edward Gifford (Catherine Elizabe...       1  64.0
115                                  Fortune, Mr. Mark       1  64.0
217                       Nicholson, Mr. Arthur Ernest       1  64.0
303                           Warren, Mr. Frank Manley       1  64.0
6                    Andrews, Miss

---

## Part 8: Chaining Operations (Track B)

Combine filtering, selecting, and sorting in one statement.

### Example: Chaining Multiple Operations

In [135]:
# Chain: Filter → Select columns → Sort → Display
result = titanic[
    titanic['survived'] == 1                    # Step 1: Filter for survivors
][['name', 'age', 'pclass']                     # Step 2: Select specific columns
].sort_values('age'                             # Step 3: Sort by age
).head(10)                                       # Step 4: Get first 10

print(result)

# This gets the 10 youngest survivors with their name, age, and class

                                         name   age  pclass
763   Dean, Miss. Elizabeth Gladys "Millvina"  0.17       3
1240          Thomas, Master. Assad Alexander  0.42       3
427                 Hamalainen, Master. Viljo  0.67       2
658             Baclini, Miss. Helene Barbara  0.75       3
657                    Baclini, Miss. Eugenie  0.75       3
611                 Aks, Master. Philip Frank  0.83       3
359             Caldwell, Master. Alden Gates  0.83       2
548           Richards, Master. George Sibley  0.83       2
590                     West, Miss. Barbara J  0.92       2
1              Allison, Master. Hudson Trevor  0.92       1


### Exercise 8.1: Female Survivors by Age

Chain operations to:
1. Filter for female survivors
2. Select name, age, pclass columns
3. Sort by age
4. Display first 10

In [136]:
# Your code here (one long chain!)
result = titanic[
    (titanic['survived'] == 1) & (titanic['sex'] == 'female')
][
    ['name', 'age', 'pclass']
].sort_values('age').head(10)
print(result)


                                         name   age  pclass
763   Dean, Miss. Elizabeth Gladys "Millvina"  0.17       3
657                    Baclini, Miss. Eugenie  0.75       3
658             Baclini, Miss. Helene Barbara  0.75       3
590                     West, Miss. Barbara J  0.92       2
1048              Nakid, Miss. Maria ("Mary")  1.00       3
895              Johnson, Miss. Eleanor Ileen  1.00       3
1187          Sandstrom, Miss. Beatrice Irene  1.00       3
478                     Laroche, Miss. Louise  1.00       2
866                  Hirvonen, Miss. Hildur E  2.00       3
540                  Quick, Miss. Phyllis May  2.00       2


### Exercise 8.2: First Class by Fare

Chain operations to:
1. Filter for first class
2. Select name, age, fare
3. Sort by fare (highest first)
4. Display first 15

In [137]:
# Your code here
result = titanic[
    titanic['pclass'] == 1 
    ][['name', 'age', 'fare'] 
    ].sort_values('fare', ascending=False 
    ).head(15) 
print(result)


                                                  name   age      fare
302                                   Ward, Miss. Anna  35.0  512.3292
49                  Cardeza, Mr. Thomas Drake Martinez  36.0  512.3292
183                             Lesurer, Mr. Gustave J  35.0  512.3292
50   Cardeza, Mrs. James Warburton Martinez (Charlo...  58.0  512.3292
115                                  Fortune, Mr. Mark  64.0  263.0000
116                Fortune, Mrs. Mark (Mary McDougald)  60.0  263.0000
111                     Fortune, Miss. Alice Elizabeth  24.0  263.0000
113                         Fortune, Miss. Mabel Helen  23.0  263.0000
112                         Fortune, Miss. Ethel Flora  28.0  263.0000
114                     Fortune, Mr. Charles Alexander  19.0  263.0000
250                         Ryerson, Miss. Emily Borie  18.0  262.3750
253    Ryerson, Mrs. Arthur Larned (Emily Maria Borie)  48.0  262.3750
35                            Bowen, Miss. Grace Scott  45.0  262.3750
66    

### Exercise 8.3: Young Male Survivors

Chain operations to:
1. Filter for males under 18 who survived
2. Select name, age, pclass
3. Sort by age (youngest first)
4. Display all

In [138]:
## Exercise 8.3: Young Male Survivors (Corrected)
result = titanic[
    # Step 1: Filter for males under 18 who survived
    (titanic['survived'] == 1) & (titanic['sex'] == 'male') & (titanic['age'] < 18)
][
    # Step 2: Select columns (This must be in its own brackets)
    ['name', 'age', 'pclass']
].sort_values(
    # Step 3: Sort by age (youngest first)
    'age'
) # Step 4: Display all (no .head() needed)

print(result)

                                                 name    age  pclass
1240                  Thomas, Master. Assad Alexander   0.42       3
427                         Hamalainen, Master. Viljo   0.67       2
548                   Richards, Master. George Sibley   0.83       2
359                     Caldwell, Master. Alden Gates   0.83       2
611                         Aks, Master. Philip Frank   0.83       3
1                      Allison, Master. Hudson Trevor   0.92       1
339                         Becker, Master. Richard F   1.00       2
492                             Mallet, Master. Andre   1.00       2
762                        Dean, Master. Bertram Vere   1.00       3
587                       Wells, Master. Ralph Lester   2.00       2
514                    Navratil, Master. Edmond Roger   2.00       2
734            Coutts, Master. William Loch "William"   3.00       3
549                    Richards, Master. William Rowe   3.00       2
515                        Navrati

---

## Part 9: Extended Statistics (Track B)

### Example: Statistics by Subgroups

In [139]:
# Calculate statistics for different subgroups

# Average fare for first class
first_class = titanic[titanic['pclass'] == 1]
avg_first_fare = first_class['fare'].mean()
print(f"First class average fare: ${avg_first_fare:.2f}")

# Average fare for third class
third_class = titanic[titanic['pclass'] == 3]
avg_third_fare = third_class['fare'].mean()
print(f"Third class average fare: ${avg_third_fare:.2f}")

# Compare the two
print(f"Difference: ${avg_first_fare - avg_third_fare:.2f}")

First class average fare: $87.51
Third class average fare: $13.30
Difference: $74.21


### Exercise 9.1: Statistics by Class

Calculate the average fare for each class separately.

Which class paid the most on average?

In [140]:
# Your code here
avg_fare_by_class = titanic.groupby('pclass')['fare'].mean()

print(avg_fare_by_class)



pclass
1    87.508992
2    21.179196
3    13.302889
Name: fare, dtype: float64


### Exercise 9.2: Survival Rates by Gender

Calculate the percentage of males who survived and the percentage of females who survived.

Which gender had a higher survival rate?

In [141]:
# Your code here
survival_rates = titanic.groupby('sex')['survived'].mean() * 100

print("Survival Rates (%):")
print(survival_rates)

# Females had a higher survival rate (approximately 74.2% vs 18.9%).


Survival Rates (%):
sex
female    72.746781
male      19.098458
Name: survived, dtype: float64


### Exercise 9.3: Age Statistics by Survival

Calculate:
1. Average age of survivors
2. Average age of non-survivors
3. Who was older on average?

In [142]:
# Your code here

avg_age_by_survival = titanic.groupby('survived')['age'].mean()

print("Average Age by Survival (0=Non-Survivor, 1=Survivor):")
print(avg_age_by_survival)



Average Age by Survival (0=Non-Survivor, 1=Survivor):
survived
0    30.545363
1    28.918244
Name: age, dtype: float64


---

## Part 10: Analysis Questions

### Question 1: Survival Rate

What percentage of passengers survived?

In [143]:
# Your code here
survival_percentage = titanic['survived'].mean() * 100
print(f"Percentage of passengers who survived: {survival_percentage:.2f}%")



Percentage of passengers who survived: 38.20%


**Answer:** 

### Question 2: Class Distribution

Which class had the most passengers?

In [144]:
# Your code here
class_distribution = titanic['pclass'].value_counts()
print(class_distribution)

# Answer:
# Class 3 had the most passengers.


pclass
3    709
1    323
2    277
Name: count, dtype: int64


**Answer:** 

### Question 3: Age Comparison

What was the average age of first class vs third class passengers?

In [145]:
# Your code here

class_age_comparison = titanic[
    (titanic['pclass'] == 1) | (titanic['pclass'] == 3)
].groupby('pclass')['age'].mean()

print(class_age_comparison)

# Answer:
# First class passengers were older on average (approx. 39 years) compared to 
# third class passengers (approx. 24 years).

pclass
1    39.159930
3    24.816367
Name: age, dtype: float64


**Answer:** 

### Question 4: Gender Distribution

Were there more male or female passengers?

In [146]:
# Your code here
gender_counts = titanic['sex'].value_counts()
print(gender_counts)

sex
male      843
female    466
Name: count, dtype: int64


**Answer:** 

### Question 5: Fare Comparison

Did survivors pay more on average than non-survivors?

In [147]:
# Your code here
avg_fare_by_survival = titanic.groupby('survived')['fare'].mean()
print(avg_fare_by_survival)

survived
0    23.353831
1    49.361184
Name: fare, dtype: float64


**Answer:** 

### Question 6: Child Survival Rate

What percentage of children (age <= 12) survived compared to adults (age > 12)?

In [148]:
# Your code here

titanic['is_child'] = titanic['age'] <= 12


child_survival_rate = titanic.groupby('is_child')['survived'].mean() * 100
print(child_survival_rate)

is_child
False    36.707819
True     57.446809
Name: survived, dtype: float64


**Answer:** 

### Question 7: Fare and Survival

Did survivors pay higher fares on average than non-survivors? Why might this be?

In [149]:
# Your code here

avg_fare_by_survival = titanic.groupby('survived')['fare'].mean()
print(avg_fare_by_survival)

survived
0    23.353831
1    49.361184
Name: fare, dtype: float64


**Answer:** 

### Question 8: Class Survival Rates

Calculate the survival rate for each class (1st, 2nd, 3rd). Which class had the highest survival rate?

In [150]:
# Your code here
class_survival_rates = titanic.groupby('pclass')['survived'].mean() * 100
print(class_survival_rates)


pclass
1    61.919505
2    42.960289
3    25.528914
Name: survived, dtype: float64


**Answer:** pclass
1    61.919503
2    42.960299
3    25.539823

---

## Part 11: Critical Thinking (Track B)

### Question 9: Important Factors

Based on your analysis, what were the THREE most important factors for survival?

Support with specific statistics.

**Your Answer:**
Gender (Female): Survival rate for females was much higher at ~74.2% vs. males at ~18.9%.

Class (First Class): First class passengers had the best chance of survival at ~61.9% (vs. 3rd class at ~25.5%).

Age (Younger): The average age of survivors (~28.9 years) was lower than non-survivors (~31.1 years).

### Question 10: Pandas to SQL

You filtered with: `titanic[(titanic['sex'] == 'female') & (titanic['survived'] == 1)]`

How do you think this might look in SQL? (Guess - we'll learn this next!)

**Your Answer:** SELECT * FROM titanic 
WHERE sex = 'female' AND survived = 1;

---

## Submission Checklist

- [ ] Completed all exercises including Track B sections
- [ ] Run all cells successfully
- [ ] Added name and date at top
- [ ] Answered all questions
- [ ] Saved as `dbAppsTask02TrackB.ipynb`
- [ ] Pushed to GitHub

**Excellent work! You're ready for SQL!**